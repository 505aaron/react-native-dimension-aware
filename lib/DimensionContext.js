import React, { Component, createContext } from "react";
import PropTypes from "prop-types";
import { Dimensions } from "react-native";
import reducer, { dimensionHydrate, dimensionChanged } from "./reducer";

const { Provider, Consumer } = createContext(reducer({}, dimensionHydrate()));

export class DimensionProvider extends Component {
  static propTypes = {
    children: PropTypes.oneOfType([
      PropTypes.node,
      PropTypes.arrayOf(PropTypes.node)
    ])
  };

  constructor(props) {
    super(props);

    this.state = {
      data: reducer({}, dimensionHydrate())
    };
  }

  componentDidMount() {
    Dimensions.addEventListener("change", this.dimensionHandler);
  }

  componentWillUnmount() {
    Dimensions.removeEventListener("change", this.dimensionHandler);
  }

  dimensionHandler = ({ window, screen }) => {
    this.setState(({ data }) => ({
      data: reducer(data, dimensionChanged(window, screen))
    }));
  };

  render() {
    const { children } = this.props;
    const { data } = this.state;

    return <Provider value={data}>{children}</Provider>;
  }
}

export default Consumer;
