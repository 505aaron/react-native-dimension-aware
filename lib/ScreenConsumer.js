import React from "react";
import PropTypes from "prop-types";
import DimensionConsumer from "./DimensionContext";
import { getScreenWidth, getScreenHeight } from "./selectors";

const ScreenConsumer = ({ children }) => {
  return (
    <DimensionConsumer>
      {dimensions =>
        children(getScreenWidth(dimensions), getScreenHeight(dimensions))
      }
    </DimensionConsumer>
  );
};

ScreenConsumer.propTypes = {
  children: PropTypes.func.isRequired
};

export default ScreenConsumer;
