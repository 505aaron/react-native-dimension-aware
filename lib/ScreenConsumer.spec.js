import React from "react";
import { View, Dimensions, Text } from "react-native";
import TestRenderer from "react-test-renderer";
import { DimensionProvider } from "./DimensionContext";
import ScreenConsumer from "./ScreenConsumer";

describe("ScreenConsumer", () => {
  let component;

  beforeEach(() => {
    component = (
      <DimensionProvider>
        <ScreenConsumer>
          {(width, height) => (
            <View>
              <Text>Width {width}</Text>
              <Text>Height {height}</Text>
            </View>
          )}
        </ScreenConsumer>
      </DimensionProvider>
    );
  });

  describe("lifecyle", () => {
    test("render context", () => {
      Dimensions.set({ screen: { width: 200, height: 200 } });

      const testRenderer = TestRenderer.create(component);

      expect(testRenderer.toJSON()).toMatchSnapshot();
    });

    test("should re-render the component on updates", () => {
      const testRenderer = TestRenderer.create(component);

      Dimensions.set({ screen: { width: 4000, height: 2000 } });

      testRenderer.update(component);

      expect(testRenderer.toJSON()).toMatchSnapshot();
    });

    test("should remove listeners", () => {
      const testRenderer = TestRenderer.create(component);

      testRenderer.unmount();

      Dimensions.set({ screen: { width: 4000, height: 2000 } });

      expect(testRenderer.toJSON()).toBeNull();
    });
  });
});
