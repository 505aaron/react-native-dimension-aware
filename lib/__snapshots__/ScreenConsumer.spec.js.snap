// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ScreenConsumer lifecyle render context 1`] = `
<View>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
  >
    Width 
    200
  </Text>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
  >
    Height 
    200
  </Text>
</View>
`;

exports[`ScreenConsumer lifecyle should re-render the component on updates 1`] = `
<View>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
  >
    Width 
    4000
  </Text>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
  >
    Height 
    2000
  </Text>
</View>
`;
